File: backend/src/models/Transaction.ts
import mongoose, { Schema, Document } from 'mongoose';
import { ISubscription } from './Subscription';
import { IUser } from './User';

export interface ITransaction extends Document {
  subscriptionId: ISubscription['_id'];
  ownerId: IUser['_id']; // The user who owns the subscription (e.g., Alex)
  amountKes: number;
  status: 'PENDING' | 'SUCCESS' | 'FAILED';
  mpesaReceiptNo?: string;
  darajaRequestId: string;
  retryCount: number;
  transactionDate: Date;
}

const TransactionSchema: Schema = new Schema({
  subscriptionId: { type: Schema.Types.ObjectId, ref: 'Subscription' },
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  amountKes: { type: Number, required: true },
  status: { type: String, enum: ['PENDING', 'SUCCESS', 'FAILED'], default: 'PENDING' },
  mpesaReceiptNo: { type: String },
  darajaRequestId: { type: String, required: true, unique: true },
  retryCount: { type: Number, default: 0 },
  transactionDate: { type: Date, default: Date.now },
}, { timestamps: true });

TransactionSchema.index({ ownerId: 1 });
TransactionSchema.index({ darajaRequestId: 1 });

export default mongoose.model<ITransaction>('Transaction', TransactionSchema);

File: backend/app/models/transaction.py
from datetime import datetime, timezone
from typing import Optional, Literal, Any
from beanie import Document, Link, PydanticObjectId
from pydantic import Field


class Transaction(Document):
    subscription_id: Optional[PydanticObjectId] = Field(
        default=None, alias="subscriptionId"
    )
    owner_id: PydanticObjectId = Field(alias="ownerId")
    amount_kes: float = Field(alias="amountKes")
    status: Literal["PENDING", "SUCCESS", "FAILED"] = "PENDING"
    mpesa_receipt_no: Optional[str] = Field(default=None, alias="mpesaReceiptNo")
    daraja_request_id: str = Field(alias="darajaRequestId")
    retry_count: int = Field(default=0, alias="retryCount")
    transaction_date: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="transactionDate"
    )
    callback_data: Optional[dict[str, Any]] = Field(default=None, alias="callbackData")
    failure_reason: Optional[str] = Field(default=None, alias="failureReason")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "transactions"
        indexes = [
            "owner_id",
            "daraja_request_id",
        ]
